core/src/org/sbml/jsbml/util/Maths.java

```diff
@@ -296,10 +296,21 @@ public static final double log(double number, double base) {
    * @return the rootExponent-th root of the radiant
    */
   public static final double root(double radiant, double rootExponent) {
-    if (rootExponent != 0) {
-      return Math.pow(radiant, 1 / rootExponent);
+    if (rootExponent == 0d) {
+      throw new ArithmeticException("Root exponent must not be zero.");
+    }
+    else if(rootExponent == 1d) {
```

Comment:
I don't like this large additional complexity to introduce a micro-optimization. 
This is basically the fallacy of premature optimization in action. "premature optimization is the root of all evil."
Did anybody complain this being too slow? There are other places to start. No optimization without benchmarking. I would revert this to the simple readible one-liner on the left. 
Probably it is even slower then before because the function cannot be inlined any more ;/

Reply:
We should possibly benchmark it using SBSCL.

Reply:
@matthiaskoenig The idea of this change was not improve the runtime but to improve the accuracy of the results as `Math.pow()` doesn't seem to use `Math.cbrt()` internally if called with 1/3 as the exponent. Example: If i wanted to calculate what the third root of 64 is `pow(64, 1d/3d)` would return 3.9999999999999996 while `cbrt(64)` would return 4.0 . Maybe that wasn't quite clear from my description. Nevertheless we could of course still benchmark it if desired.

Reply:
Okay, I get it. But still don't like it. The problem you could get now is that you have possible different error accumulation depending on the exponents. I.e. if you do multiple merges with 1/3 you will have different error accumulation then with 1/2.

I still think this is overkill and in my opinion not needed. This is math on doubles and floating point errors are to be expected. The point that I by just reading the code expected the reasons for the different cases is a performance issue already shows that this is unclear and results in confusion. So if this is required please add documentation to the code for the reasons why these switch is there.

Reply:
@matthiaskoenig Thank you! I am really glad you are so careful and conservative here. I think, the main use case for this specific function is merging units. There you want to have cancellation and sometimes need to come close to integer values if that is possible. Java does already provide specific functions for this, all that is done here is a case distinction followed by redirection to existing built-in functions of Java.

Reply:
So I have now performed runtime analysis between the different versions of the `root()` method with `Original (König)` being the original version as it was before and `New 1 (Pertuch)` and `New 2 (Dräger)` the new versions without and with the requested changes by @draeger. Here a test data set was created containing 100 million values where every tenth value was a square root and every other tenth value was a cubic root. The rest of the values were random. Then the root function was called for each of these values either always with the same root exponent for each value (1, 2, 3, 4, 8 and 20) or with a list of root exponents for each test data value which was either completely random or one where 2s and 3s were fixed at the same position as the square/cubic roots in the test data.  The results can be seen in first of the attached figures. The original version always takes 5.7-5.9s for the complete data set. The new versions are much faster for root exponents of 1 and 2 and slightly faster for a root exponent of 3. For root exponents >4 they perform roughly the same. For the list of random exponents all algorithms perform approximately the same while the new versions are faster if some values were fixed to 2 or 3. 

Besides runtime also the average difference between the results of the original and the new algorithms was examined with the same setup as described above. The results can be seen in the second figure where only the differences between the original and the new algorithm is displayed as both new `root()` methods achieve the same results. As one would expect there was only a differences if root exponents of either 2 or 3 were used but the differences were rather small with ~10^-14 and are therefore numerically negligible.

In conclusion I would suggest that we use the `root()` method from my pull request with the changes requested by @draeger as the new method has a better runtime and can produce better results if the root exponent is either 2 or 3. I could then, as @matthiaskoenig requested, describe in the JavaDoc why the method was changed. 

![runtimes_root](https://user-images.githubusercontent.com/39677786/123421581-4f3f0f00-d5bd-11eb-90bc-50c524eb89c0.png)
![differences_root](https://user-images.githubusercontent.com/39677786/123421584-50703c00-d5bd-11eb-9b30-31365b4b6902.png)

Reply:
Great thanks. All my issues have been adressed. Thanks for the runtimes, this is very interesting.